### PR TITLE
Add Steam ID implementation

### LINKED ISSUE #1: Steam.Common - Steam ID API
A [Steam ID](https://developer.valvesoftware.com/wiki/SteamID) is a globally unique identifier for Steam accounts, groups, lobbies, and chat rooms.

Steam IDs are used throughout Steam, on the web and in the client. They have three formats:

 * Raw bits: 64 bits in length, sometimes called the "community ID"
 * Steam2: a common format for public user IDs
 * Steam3: a format used in debugging and games as they can support a larger variety of account types

The `SteamId` type is a read-only struct that represents the data stored in these formats. It can handle conversions from a raw values, Steam2 format strings, and Steam3 format string. It also allows retrieval of the individual pieces of information a Steam ID makes up, including account universe, type, ID, and instance.

Three other enums will need to be added to support `SteamId`: `Universe`, `AccountType`, and `Instance`.

### Details

Constructors for `SteamId` check that the ID is valid, so all `SteamId` values must be valid. Unchecked constructors may be added in the future.

`SteamId` has a number of predefined values defined in the public client header as part of the Steamworks SDK. These predefined values are listed in the API outline below. These predefined values also contain `Zero` for the default `SteamId` represented in raw as just zero.

Properties are exposed for getting the universe, type, ID, and instance of the account this `SteamId` is for.

Two constructors are provided. The first constructor is for individually setting each property of the `SteamId` (instance is range checked as it's 20 bits long and throws an out of range exception when invalid). The second constructor is for creating a individual user `SteamId` with the specified instance flags (instance is also range checked and throws an out of range exception when invalid).

Two methods are borrowed from the public client header that creates anonymous user and anonymous game server IDs. Check the Steamworks SDK for info about implementing these methods.

Three methods are defined for converting from the three `ID` formats. `FromUInt64`, `FromSteam2`, and `FromSteam3` convert from their respective formats as necessary. Invalid inputs throw `ArgumentException`.

The `Is` properties after are validity checks and simple properties pulled from public client headers. Implementations for these can be ported from there.

`ToChat`, `ToClan`, `ClanToChat`, and `ChatToClan` are also ported from client headers.

`ToStaticAccountKey` is also ported from client headers. The other `To` methods are counterparts to their `From` methods, throwing `InvalidOperationException` on incompatible IDs instead of `ArgumentException`.

`SteamId` overrides all object methods and implements `IEquatable<SteamId>`. `ToString` returns the ID in Steam3 format. It also includes a `StaticEquals` method for comparing the static properties of the IDs. Implicit and explicit operators are provided for converting to and from raw bits.

### API

```cs
public readonly struct SteamId : IEquatable<SteamId>
{
    static readonly SteamId Zero;

    // predefined from sdk
    static readonly SteamId OutOfDateGameServer;
    static readonly SteamId LanModeGameServer;
    static readonly SteamId NetYetInitializedGameServer;
    static readonly SteamId NoSteamGameServer;

    readonly ulong Value;

    Universe AccountUniverse { get; }
    AccountType AccountType { get; }
    uint AccountId { get; }
    uint AccountInstance { get; }

    SteamId(ulong id);
    SteamId(uint id, AccountType type, Universe universe, uint instance);

    static SteamId CreateAnonymousUser(Universe universe);
    static SteamId CreateAnonymousGameServer(Universe universe);
    static SteamId CreateIndividualUser(uint id, Universe universe, Instance instance);

    static SteamId FromSteam2(string id);
    static SteamId FromSteam3(string id);

    static bool IsValid(SteamId id);
    bool IsBlankAnonymousAccount { get; }
    bool IsPersistentGameServer { get; }
    bool IsGameServer { get; }
    bool IsContentServer { get; }
    bool IsClan { get; }
    bool IsChat { get; }
    bool IsIndividualAccount { get; }
    bool IsConsoleUser { get; }
    bool IsAnonymousGameServer { get; }
    bool IsAnonymousUser { get; }
    bool IsAnonymousAccount { get; }
    bool IsLobby { get; }
    
    static SteamId ToChat(SteamId id);
    static SteamId ToClan(SteamId id);
    static SteamId ChatToClan(SteamId id);
    static SteamId ClanToChat(SteamId id);

    static ulong ToStaticAccountKey(SteamId id);
    static string ToSteam2(SteamId id);
    static string ToSteam3(SteamId id);

    bool StaticEquals(SteamId other);
    bool Equals(SteamId id);
    override bool Equals(object obj);
    override bool GetHashCode();
    override string ToString();
    static implicit operator ulong(SteamId id);
    static implicit operator SteamId(ulong id);
    static bool operator ==(SteamId a, SteamId b);
    static bool operator !=(SteamId a, SteamId b);
}

public enum Universe
{
    Invalid,
    Public,
    Beta,
    Internal,
    Dev
}

public enum AccountType
{
    Invalid,
    Individual,
    Multiseat,
    GameServer,
    AnonGameServer,
    Pending,
    ContentServer,
    Clan,
    Chat,
    ConsoleUser,
    AnonUser
}

[Flags]
public enum Instance
{
    All = 0,
    Desktop = 1,
    Console = 2,
    Web = 4
}
```

# Changes

* ID was changed to Id by popular vote.
* Many methods were made static as to not look like they change state.
* Pending and Unknown tracking was removed.
* ~~What constitutes valid input has been more constrained.~~
* Moved individual account constructor to static helper method.
* Revert decision on constrained valid input.
* Move FromUInt64 to constructor. 
* Remove ToUInt64.
* Remove IsGroupChat since it was removed from public headers
* Expose underlying value

### PR DESCRIPTION
**Pull request by ObsidianMinor**
Closes #1

### CODE REVIEW

File: src/Steam.Common/SteamId.cs [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+                    return false;
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Returns whether this is a login ID to be filled in
+        /// </summary>
+        public bool IsBlankAnonymousAccount => AccountId == 0 && IsAnonymousAccount && AccountInstance == 0;
```
Comment (JurjenBiewenga) [inline src/Steam.Common/SteamId.cs]: I'd like to see one white line between properties to make the distinction more clear.

File: src/Steam.Common/SteamId.cs [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+using System.Text.RegularExpressions;
+
+namespace Steam
+{
+    /// <summary>
+    /// A globally unique identifier for Steam accounts, groups, lobbies, and chat rooms
+    /// </summary>
+    public readonly struct SteamId : IEquatable<SteamId>
+    {
+        private const uint AccountIdMask = 0xFFFFFFFF;
```
Comment (JurjenBiewenga) [inline src/Steam.Common/SteamId.cs]: We should consider whether we want to add documentation to private members as well.

File: src/Steam.Common/SteamId.cs [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+        private const int AccountInstanceShift = 32;
+
+        private const uint AccountTypeMask = 0xF;
+        private const int AccountTypeShift = 52;
+
+        private const int AccountUniverseShift = 56;
+
+        private const int ChatAccountInstanceMask = 0xFFF;
+
+        private readonly static Regex Steam2Regex = new Regex(@"STEAM_([0-4]):(0|1):(\d+)");
```
Comment (JurjenBiewenga) [inline src/Steam.Common/SteamId.cs]: The order of these modifiers is inconsistent with the rest. We should pick one way and stick with it.

File: src/Steam.Common/AccountType.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
+﻿namespace Steam
```
Comment (JurjenBiewenga) [inline src/Steam.Common/AccountType.cs]: Shouldn't this and the other namespaces be `Steam.Common`?

File: tests/Steam.Common.Tests/SteamIdTests.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
+﻿using System;
+using Xunit;
+
+namespace Steam.Tests
+{
+    public class SteamIdTests
```
Comment (JurjenBiewenga) [inline tests/Steam.Common.Tests/SteamIdTests.cs]: I'd like to see more tests to check the correctness of equality operators and preferably also the properties to make sure we can't accidentally break these.

File: src/Steam.Common/SteamId.cs [hunk 6]
```diff
@@ -0,0 +1,492 @@
+﻿using System;
+using System.Text.RegularExpressions;
+
+namespace Steam
+{
+    /// <summary>
+    /// A globally unique identifier for Steam accounts, groups, lobbies, and chat rooms
+    /// </summary>
+    public readonly struct SteamId : IEquatable<SteamId>
```
Comment (JurjenBiewenga) [inline src/Steam.Common/SteamId.cs]: I don't think it is a bad thing to link to the Steam API for more information about this construct, if we want to do this we should keep it consistent in the whole project (Currently app, depot, and package do have this)

File: src/Steam.Common/SteamId.cs [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+        /// 
+        /// For all other account types, the instance is Z or 0 if Z wasn't found.
+        /// </remarks>
+        public static SteamId FromSteam3(string id)
+        {
+            var match = Steam3Regex.Match(id);
+            if (!match.Success)
+                throw new ArgumentException("The input string does not match the Steam3 pattern");
+
+            var typeChar = match.Groups[1].Value[0];
```
Comment (JurjenBiewenga) [inline src/Steam.Common/SteamId.cs]: Nice use of groupings, however using named groups might make the code more readable

**Review by JurjenBiewenga — CHANGES_REQUESTED** [review]
Looks pretty good, mostly some stylistic issues. Unfortunately, I haven't been able to fact check your implementation against the steam api.

File: src/Steam.Common/AccountType.cs (hunk 4 above)
Comment (ObsidianMinor) [reply src/Steam.Common/AccountType.cs]: Was your stuff in Steam.Common?

File: src/Steam.Common/SteamId.cs (hunk 6 above)
Comment (ObsidianMinor) [reply src/Steam.Common/SteamId.cs]: The Steam API docs are those docs
image

File: src/Steam.Common/SteamId.cs (hunk 7 above)
Comment (ObsidianMinor) [reply src/Steam.Common/SteamId.cs]: With constant ints?

File: src/Steam.Common/AccountType.cs (hunk 4 above)
Comment (JurjenBiewenga) [reply src/Steam.Common/AccountType.cs]: Yeah it is, feel free to change it.

File: src/Steam.Common/SteamId.cs (hunk 6 above)
Comment (JurjenBiewenga) [reply src/Steam.Common/SteamId.cs]: That's all there is that's documented on their site? If it is then that it is fine like this

File: src/Steam.Common/AccountType.cs (hunk 4 above)
Comment (ObsidianMinor) [reply src/Steam.Common/AccountType.cs]: Nope, I checked, it's in `Steam` (I set the root namespace ahead of time).

File: src/Steam.Common/SteamId.cs (hunk 7 above)
Comment (JurjenBiewenga) [reply src/Steam.Common/SteamId.cs]: Nah, you can add names to groups in regex. Lke this https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named-matched-subexpressions

File: src/Steam.Common/SteamId.cs (hunk 2 above)
Comment (ObsidianMinor) [reply src/Steam.Common/SteamId.cs]: I don't think it's necessary, unless the private member is incredibly vague we may add a comment but I don't think we should add too many docs.

File: src/Steam.Common/SteamId.cs (hunk 3 above)
Comment (ObsidianMinor) [reply src/Steam.Common/SteamId.cs]: I'd say `static` should be followed by `readonly`. I can't figure out any way to enforce this however.

File: tests/Steam.Common.Tests/SteamIdTests.cs (hunk 5 above)
Comment (ObsidianMinor) [reply tests/Steam.Common.Tests/SteamIdTests.cs]: I believe the "parts" tests test these well.

**Pull request merged**